### PR TITLE
Core/Porting OMP loops to Parallel Utilities-IV

### PR DESCRIPTION
**Pull request by ashishdarekar**
**Description**
Porting of OMP Parallel For loops to Parallel Utilities

Please mark the PR with appropriate tags: 
- ParallelUtilities
- AtomicUtilities

**Changelog**
-Replaced OMP for loop with Private variables with Parallel Utilities with TLS
-Replaced OMP Atomic with Atomic Utilities
-Replaced OMP for loop with Variable_Utils

### CODE REVIEW

File: kratos/processes/find_nodal_h_process.cpp [hunk 1]
```diff
@@ -34,11 +35,14 @@ void FindNodalHProcess<THistorical>::Execute()
...
-        auto it_node = mrModelPart.NodesBegin() + i;
-        SetInitialValue(it_node);
-    }
+
+    //Using Parallel Utilities
+    IndexPartition<std::size_t>(mrModelPart.Nodes().size()).for_each([&](std::size_t index){
+        SetInitialValue(mrModelPart.NodesBegin() + index);
+    });
+
+    //Or Using Variable_utils
```
Comment (ashishdarekar) [inline kratos/processes/find_nodal_h_process.cpp]: @philbucher can we use Variable utils here??? Compilation was successful, but whether will it work always ??
SetInitialValue() here is similar to SetVariable from Variable Utils

File: kratos/processes/compute_nodal_normal_divergence_process.cpp [hunk 2]
```diff
@@ -27,10 +27,12 @@ void ComputeNodalNormalDivergenceProcess<THistorical>::Execute()
     ClearDivergence();
 
     // Auxiliary containers
-    Matrix J0, InvJ0, DN_DX;
+    struct tls_type
```
Comment (ashishdarekar) [inline kratos/processes/compute_nodal_normal_divergence_process.cpp]: Defined Struct Here only...

File: kratos/processes/compute_nodal_gradient_process.cpp [hunk 3]
```diff
@@ -81,60 +81,70 @@ void ComputeNodalGradientProcess<THistorical>::ComputeElementalContributionsAndV
         p_variable_retriever = new VariableVectorRetriever<ComputeNodalGradientProcessSettings::GetAsHistoricalVariable>();
     }
 
+    struct tls_type
```
Comment (ashishdarekar) [inline kratos/processes/compute_nodal_gradient_process.cpp]: Redefining those variables again, Is it okay?? @philbucher

File: kratos/processes/compute_nodal_normal_divergence_process.cpp (hunk 2 above)
Comment (philbucher) [reply kratos/processes/compute_nodal_normal_divergence_process.cpp]: perfect!

File: kratos/processes/find_nodal_h_process.cpp (hunk 1 above)
Comment (philbucher) [reply kratos/processes/find_nodal_h_process.cpp]: no we can not, `SetInitialValue()` behaves differently depending on the class template

File: kratos/processes/compute_nodal_gradient_process.cpp [hunk 4]
```diff
@@ -74,69 +74,77 @@ void ComputeNodalGradientProcess<THistorical>::ComputeElementalContributionsAndV
...
+        Vector N, values;
+        double detJ0 = 0.0;
+    };
+    tls_type TLS = tls_type();
+    TLS.DN_DX = DN_DX;
+    TLS.J0 =J0;
+    TLS.InvJ0 = InvJ0;
+    TLS.N = N;
+    TLS.values = values;
+    TLS.detJ0 = detJ0;
```
Comment (philbucher) [inline kratos/processes/compute_nodal_gradient_process.cpp]: you can directly use the tls in the initialization that happens above, then you don't need to do the copy here

**Review by philbucher — COMMENTED** [review]
nice!

File: kratos/processes/find_nodal_h_process.cpp (hunk 1 above)
Comment (ashishdarekar) [reply kratos/processes/find_nodal_h_process.cpp]: Got it .

File: kratos/processes/compute_nodal_gradient_process.cpp [hunk 5]
```diff
@@ -50,9 +50,13 @@ template<bool THistorical>
...
-    Matrix DN_DX, J0, InvJ0;
-    Vector N, values;
-    double detJ0 = 0.0;
+    struct tls_type
+    {
+        Matrix DN_DX, J0, InvJ0;
+        Vector N, values;
+        double detJ0 = 0.0;
+    };
+    tls_type TLS = tls_type();
```
Comment (philbucher) [inline kratos/processes/compute_nodal_gradient_process.cpp]:
```suggestion
    tls_type tls;
```
following the naming convention

File: kratos/processes/compute_nodal_gradient_process.cpp [hunk 6]
```diff
@@ -50,9 +50,13 @@ template<bool THistorical>
 void ComputeNodalGradientProcess<THistorical>::ComputeElementalContributionsAndVolume() {
 
     // Auxiliar containers
-    Matrix DN_DX, J0, InvJ0;
-    Vector N, values;
-    double detJ0 = 0.0;
+    struct tls_type
```
Comment (philbucher) [inline kratos/processes/compute_nodal_gradient_process.cpp]:
```suggestion
    struct TLSType
```

**Review by philbucher — APPROVED** [review]
perfect :+1:

**Comment by ashishdarekar** [comment]
> perfect

Why CI is failing ?? Is there any other problem?

**Comment by philbucher** [comment]
> > perfect
> 
> Why CI is failing ?? Is there any other problem?

this Build is new, there are some problems with it

you only need to concern yourself with the checks that have the "required" label

Also please always run the tests locally, since the CI only runs a subset of tests!

**Pull request merged**